content/posts/2024-11-05-flang-new.md

```diff
@@ -0,0 +1,792 @@
+`arith` (arithmetic) dialect as well.
+
+```mlir
+module attributes {<...>} {
+  func.func @_QPfoo(<...>) {
+    <...>
+    %c3_i32 = arith.constant 3 : i32
+    %7 = fir.convert %0 : (!fir.ref<!fir.box<!fir.array<?xf32>>>) -> !fir.ref<!fir.box<none>>
+    %8 = fir.convert %5 : (!fir.box<!fir.array<?xf32>>) -> !fir.box<none>
+    %9 = fir.convert %6 : (!fir.ref<!fir.char<1,17>>) -> !fir.ref<i8>
+    %10 = fir.call @_FortranAAssign(%7, %8, %9, %c3_i32) : (!fir.ref<!fir.box<none>>, !fir.box<none>, !fir.ref<i8>, i32) -> none
+    return
+  }
+<...>
+}
+```
+Then this bundle of MLIR dialects is converted into purely LLVM IR:
+```mlir
+define void @foo_(ptr %0, ptr %1) {
+  <...>
+  store { ptr, i64, i32, i8, i8, i8, i8, [1 x [3 x i64]] } %37, ptr %3, align 8
+  call void @llvm.memcpy.p0.p0.i32(ptr %5, ptr %4, i32 48, i1 false)
+  %38 = call {} @_FortranAAssign(ptr %5, ptr %3, ptr @_QQclX2F6170702F6578616D706C652E66393000, i32 3)
+  ret void
+}
+<...>
+```
+This LLVM IR passes through the standard optimisation pipeline that LLVM IR produced
+by Clang would. Eventually being converted into target specific Machine IR
+(MIR) and finally machine code.
+
+At each of the stages, the optimisations most suited to that stage are done.
+For example, while we are in HLFIR we could optimise array accesses because in
+HLFIR we have information about how the Fortran language treats those.
+
+If we were to do this later in LLVM IR it would be much more difficult. Either
+the information would be lost or incomplete, or we would be at a stage in the pipeline
+where we cannot assume that we started with a specific source language.
+
+# OpenMP to Everyone
+
+**Note:** Most of the points made in this section also apply to [OpenACC](https://www.openacc.org/) support for Flang. In the interest of time, I will only talk
+about OpenMP in this article. You can find out more about OpenACC support in this
+[presentation](https://www.youtube.com/watch?v=vVmCLdSboWc).
+
+## The Basics
+
+OpenMP is a standardised API for adding parallelisation to C, C++ and
+Fortran programs. Programmers mark parts of their code with directives.
+The compiler interprets these directives and produces calls to a runtime library
+that distributes work as instructed.
+
+Here is an example in Fortran:
+```Fortran
+SUBROUTINE SIMPLE(N, A, B)
+  INTEGER I, N
+  REAL B(N), A(N)
+!$OMP PARALLEL DO
+  DO I=2,N
+    B(I) = (A(I) + A(I-1)) / 2.0
+  ENDDO
+END SUBROUTINE SIMPLE
+```
+(taken from ["OpenMP Application Programming Interface Examples"](https://www.openmp.org/wp-content/uploads/openmp-examples-4.5.0.pdf), [Compiler Explorer](https://godbolt.org/z/chjzs3o6r))
+
+**Note:** Fortran arrays are [one-based](https://fortran-lang.org/en/learn/quickstart/arrays_strings/) by default. So the first element is at index 1. This example reads the previous element, so it starts `I` at 2.
+
+`!$OMP PARALLEL DO` is a directive in the form of a Fortran comment, which
```

Comment:
```suggestion
`!$OMP PARALLEL DO` is a combined directive in the form of a Fortran comment, which
```

Reply:
A combined directive is a subset of directives is what I'd argue here. From a reader's point of view not the OpenMP standard's perspective.

Unless you think not including that will make it misleading.

Reply:
Agreed.